4단계 - 수강신청(요구사항 변경)

Denia-park opened the pull request:
## 진행 내용
- 변경된 기능 요구사항에 맞춰서 내용 수정

## 질문사항
- 3단계 피드백 주신 내용 관련해서 코멘트로 몇 가지 질문 드렸는데 답변 해주실 수 있으실까요 ?? ([링크](https://github.com/next-step/java-lms/pull/302))

Denia-park committed:
docs: 피드백 사항 추가

Denia-park committed:
docs: step4 학습목표 및 요구사항 추가

Denia-park committed:
feat: Session에 생성자, 생성일, 수정자, 수정일을 추가하자.

Denia-park committed:
refact: SessionStatus에 진행중 인스턴스 추가

Denia-park committed:
refact: Enrollment에 모집 상태를 나타내는 RecruitingStatus 추가

Denia-park committed:
refact: RecruitingStatus 관련 내용들을 DB에 저장하도록 수정

Denia-park committed:
refact: 강의 진행 상태 및 모집 상태를 분리

Denia-park committed:
refact: 강의는 하나 이상의 커버 이미지를 가질 수 있다.

Denia-park committed:
feat: 선발된 유저인지 아닌지 판단하기 위해 클래스 추가

Denia-park committed:
feat: SelectionUser 관련 schema 및 Repository 생성

Denia-park committed:
refact: SelectionUsers 일급 콜렉션 추가

Denia-park committed:
refact: 기존 메서드 대신에 신규 메서드 적용

Denia-park committed:
refact: 기존 메서드 제거 후 이름 변경

Denia-park committed:
refact: 기존의 users를 사용하는 부분들을 SelectionUsers로 변경 후 기존 users 삭제

Denia-park committed:
feat: 수강신청한 유저에 대해서 선발이 되었는지, 되지 않았는지에 대한 상태변경 메서드 추가

serverwizard commented:
```
실무에서는 FK를 잘 사용하지 않는다는 이야기는 종종 들었습니다!

그럼 리뷰어님께서는 JPA로 구성할때도 서로 연관관계는 따로 맺어주시지 않고, 필요한 객체가 있으면 그때마다 따로 조회를 해서 사용하시는 걸까요 ??
```
저도 JPA로 엔티티를 구성할 때 연관관계를 맺어줘요. 다만, 연관관계를 맺었다고해서 FK를 필수로 걸진 않아요!
예를 들어 아래와 같이 엔티티를 구성할 경우, BBB 테이블에서 AAA테이블의 id는 가지지만 해당 컬럼을 인덱스로만 거는 편이에요~
```
public class AAA extends BaseEntity {
    @Id
    @GeneratedValue
    private Long id;
    ...
    @ToString.Exclude
    @OneToMany(cascade = CascadeType.PERSIST, mappedBy = "...", orphanRemoval = true)
    private List<BBB> datas = new ArrayList<>();
    ...
}
````

```
가격으로 유료, 무료가 구분이 되다보니 따로 필요가 없겠다 라는 생각이 들어서 해당 컬럼은 제외를 하게 되었습니다.
있는게 더 좋은 설계일까요 ?
```
설계에 정답은 없지만 유료/무료 강의를 구분해서 보여줘야 한다면, 가격을 보고 판단할 수도 있지만 별도 컬럼을 둘 수도 있을 것 같아 말씀 드렸어요. :) 그리고 유료 강의가 일정 기간동안 0원으로 이벤트를 할 수도 있을 것 같구요

serverwizard commented on src/main/java/nextstep/courses/domain/session/Enrollment.java:
```diff
@@ -9,6 +9,7 @@
 public class Enrollment {
     private Long price;
     private ChargeStatus chargeStatus;
+    private RecruitingStatus recruitingStatus;
```
강의 수강신청과 관련된 요구사항이 변경될 경우, Enrollment쪽만 작업하면 되네요! 👍

serverwizard commented on README.md:
```diff
@@ -152,3 +152,33 @@
+- 강사가 승인하지 않아도 수강 신청하는 모든 사람이 수강 가능하다.
+  - 우아한테크코스(무료), 우아한테크캠프 Pro(유료)와 같이 선발된 인원만 수강 가능해야 한다.
+    - 강사는 수강신청한 사람 중 선발된 인원에 대해서만 수강 승인이 가능해야 한다.
+    - 강사는 수강신청한 사람 중 선발되지 않은 사람은 수강을 취소할 수 있어야 한다.
```
혹시 위 요구사항은 어떤 걸 보고 판단하면 되는 걸까요?

serverwizard commented on src/main/java/nextstep/courses/domain/session/SelectionUser.java:
```diff
@@ -0,0 +1,37 @@
+
+public class SelectionUser {
+    SelectStatus status;
+    NsUser user;
```
수강신청한 사람에 대한 역할 👍 
접근 제한자를 붙여주시면 더 좋겠네요! :)

serverwizard commented on src/main/java/nextstep/courses/domain/session/SelectionUsers.java:
```diff
@@ -0,0 +1,37 @@
+        final List<SelectionUser> selectionUsers = new ArrayList<>();
+
+        for (final NsUser nsUser : nsUsers) {
+            selectionUsers.add(new SelectionUser(SelectStatus.UNDECIDED, nsUser));
```
만약, 위와 같이 구현할 경우 강의를 수강 신청한 사람이 선발된 인원인지 아닌지 구분이 어려울 것 같은데 어떠신가요?

serverwizard reviewed (changes requested):
형기님, 수강신청 4단계 미션 잘 구현해 주셨네요. 👍

간단한 코멘트 조금 남겼는데, 확인 부탁드릴게요 🙂

Denia-park committed:
docs: step4 피드백 사항 추가

Denia-park committed:
refact: SelectionUser의 Field에 접근제어자 추가

Denia-park committed:
refact: 수강 신청한 사람이 선발된 인원인지 아닌지 판단할 수 있도록 변수 추가

- List<NsUser>를 사용하는 경우, 아직 선발 결정을 하지 않은 유저들로 판단하여 UndecidedUsers로 처리함
- TotalSelectStatusUsers에서 유저의 상태에 따라 3가지의 타입으로 분류

Denia-park commented:
피드백 주신 사항들에 대해서 생각해보고 기존에 존재하던 SelectionUsers를 이용하여 코드를 수정해봤습니다.

확인 부탁드립니다🙇‍♂️🙇‍♂️

serverwizard commented on src/main/java/nextstep/courses/domain/session/TotalSelectStatusUsers.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public TotalSelectStatusUsers(List<SelectionUser> selectionUsers) {
+        final Map<SelectStatus, List<SelectionUser>> statusListMap = selectionUsers.stream()
```
TotalSelectStatusUsers의 인스턴스 변수로 `Map<SelectStatus, List<SelectionUser>> statusListMap`를 가지게 할 수도 있겠네요~ 👍

serverwizard reviewed (approved):
형기님, 마지막 미션까지 잘해주셨어요 💯
그리고 미션 진행하시느라 고생 많이하셨습니다 🎉

특히, 고민하시고 질문하신 내용도 인상 깊었습니다.🔥
